src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/Connection.cs

```diff
@@ -187,31 +157,65 @@ public void Tick(long timestamp)
                 {
                     _frame.SetBadRequestState(RequestRejectionReason.RequestTimeout);
                 }
-                
+
                 StopAsync();
             }
 
             Interlocked.Exchange(ref _lastTimestamp, timestamp);
         }
 
-        private void ApplyConnectionFilter()
+        private async Task ApplyConnectionAdaptersAsync()
         {
-            if (_filterContext.Connection != _libuvStream)
+            try
             {
-                _filteredStreamAdapter = new FilteredStreamAdapter(ConnectionId, _filterContext.Connection, Thread.Memory, Log, ThreadPool, _bufferSizeControl);
+                var rawStream = new RawStream(Input, Output);
+                var adapterContext = new ConnectionAdapterContext(rawStream);
+                var adaptedConnections = new IAdaptedConnection[_connectionAdapters.Count];
 
-                _frame.Input = _filteredStreamAdapter.SocketInput;
-                _frame.Output = _filteredStreamAdapter.SocketOutput;
+                for (var i = 0; i < _connectionAdapters.Count; i++)
+                {
+                    var adaptedConnection = await _connectionAdapters[i].OnConnectionAsync(adapterContext);
+                    adaptedConnections[i] = adaptedConnection;
+                    adapterContext = new ConnectionAdapterContext(adaptedConnection.ConnectionStream);
+                }
 
-                // Don't attempt to read input if connection has already closed.
-                // This can happen if a client opens a connection and immediately closes it.
-                _readInputTask = _socketClosedTcs.Task.Status == TaskStatus.WaitingForActivation ?
-                    _filteredStreamAdapter.ReadInputAsync() :
-                    TaskCache.CompletedTask;
-            }
+                if (adapterContext.ConnectionStream != rawStream)
+                {
+                    _filteredStream = adapterContext.ConnectionStream;
+                    _filteredStreamAdapter = new AdaptedPipeline(ConnectionId, adapterContext.ConnectionStream,
+                        Thread.Memory, Log, ThreadPool, _bufferSizeControl);
+
+                    _frame.Input = _filteredStreamAdapter.SocketInput;
+                    _frame.Output = _filteredStreamAdapter.SocketOutput;
+
+                    // Don't attempt to read input if connection has already closed.
+                    // This can happen if a client opens a connection and immediately closes it.
+                    _readInputTask = _socketClosedTcs.Task.Status == TaskStatus.WaitingForActivation
+                        ? _filteredStreamAdapter.ReadInputAsync()
+                        : TaskCache.CompletedTask;
+                }
 
-            _frame.PrepareRequest = _filterContext.PrepareRequest;
-            _frame.Start();
+                _frame.PrepareRequest = features =>
```

Comment:
You could avoid the allocation by setting the list of adaptedConnections on the frame itself and having it do this logic in Frame. It's actually a bit cleaner than assigning a delegate in Connection.